// angular-alert-module/esm5/lib/alerts.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { AlertsConfigData } from './alerts.config';
/**
 * @record
 */
export function AlertsInterface() { }
function AlertsInterface_tsickle_Closure_declarations() {
    /** @type {?} */
    AlertsInterface.prototype.alertsSubject;
    /** @type {?} */
    AlertsInterface.prototype.setMessage;
    /** @type {?} */
    AlertsInterface.prototype.getConfig;
    /** @type {?} */
    AlertsInterface.prototype.setConfig;
}
var AlertsService = /** @class */ (function () {
    function AlertsService() {
        // Observable alerts
        this.alertsSubject = new Subject();
    }
    // Service message commands
    /**
     * @param {?} message
     * @param {?} type
     * @return {?}
     */
    AlertsService.prototype.setMessage = /**
     * @param {?} message
     * @param {?} type
     * @return {?}
     */
    function (message, type) {
        this.alertsSubject.next({ message: message, type: type });
    };
    /**
     * @return {?}
     */
    AlertsService.prototype.getSubject = /**
     * @return {?}
     */
    function () {
        return this.alertsSubject.asObservable();
    };
    /**
     * @param {?} messagetype
     * @param {?} key
     * @return {?}
     */
    AlertsService.prototype.getConfig = /**
     * @param {?} messagetype
     * @param {?} key
     * @return {?}
     */
    function (messagetype, key) {
        if (typeof AlertsConfigData[messagetype] == 'undefined') {
            return AlertsConfigData["DefaultSettings"][key];
        }
        else {
            return (typeof AlertsConfigData[messagetype][key] != 'undefined') ? AlertsConfigData[messagetype][key] : AlertsConfigData["DefaultSettings"][key];
        }
    };
    /**
     * @param {?} messagetype
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    AlertsService.prototype.setConfig = /**
     * @param {?} messagetype
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    function (messagetype, key, value) {
        if (typeof AlertsConfigData[messagetype] == 'undefined') {
            AlertsConfigData[messagetype] = {};
        }
        AlertsConfigData[messagetype][key] = value;
    };
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    AlertsService.prototype.setDefaults = /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    function (key, value) {
        AlertsConfigData["DefaultSettings"][key] = value;
    };
    AlertsService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    AlertsService.ctorParameters = function () { return []; };
    return AlertsService;
}());
export { AlertsService };
function AlertsService_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    AlertsService.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    AlertsService.ctorParameters;
    /** @type {?} */
    AlertsService.prototype.alertsSubject;
}
