// angulartics2/esm2015/angulartics2On.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Input, NgModule, Renderer2, } from '@angular/core';
import { Angulartics2 } from './angulartics2-core';
let Angulartics2On = class Angulartics2On {
    constructor(elRef, angulartics2, renderer) {
        this.elRef = elRef;
        this.angulartics2 = angulartics2;
        this.renderer = renderer;
        this.angularticsProperties = {};
    }
    ngAfterContentInit() {
        this.renderer.listen(this.elRef.nativeElement, this.angulartics2On || 'click', (event) => this.eventTrack(event));
    }
    eventTrack(event) {
        const action = this.angularticsAction; // || this.inferEventName();
        const properties = Object.assign({}, this.angularticsProperties, { eventType: event.type });
        if (this.angularticsCategory) {
            properties.category = this.angularticsCategory;
        }
        if (this.angularticsLabel) {
            properties.label = this.angularticsLabel;
        }
        if (this.angularticsValue) {
            properties.value = this.angularticsValue;
        }
        this.angulartics2.eventTrack.next({
            action,
            properties,
        });
    }
};
tslib_1.__decorate([
    Input('angulartics2On'),
    tslib_1.__metadata("design:type", String)
], Angulartics2On.prototype, "angulartics2On", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], Angulartics2On.prototype, "angularticsAction", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], Angulartics2On.prototype, "angularticsCategory", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], Angulartics2On.prototype, "angularticsLabel", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], Angulartics2On.prototype, "angularticsValue", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], Angulartics2On.prototype, "angularticsProperties", void 0);
Angulartics2On = tslib_1.__decorate([
    Directive({ selector: '[angulartics2On]' }),
    tslib_1.__metadata("design:paramtypes", [ElementRef,
        Angulartics2,
        Renderer2])
], Angulartics2On);
export { Angulartics2On };
let Angulartics2OnModule = class Angulartics2OnModule {
};
Angulartics2OnModule = tslib_1.__decorate([
    NgModule({
        declarations: [Angulartics2On],
        exports: [Angulartics2On],
    })
], Angulartics2OnModule);
export { Angulartics2OnModule };
