// rxjs-tslint-rules/dist/rules/rxjsNoConnectableRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Lint = require("tslint");
var tsquery_1 = require("@phenomnomnominal/tsquery");
var util_1 = require("../support/util");
var Rule = (function (_super) {
    tslib_1.__extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.applyWithProgram = function (sourceFile, program) {
        var _this = this;
        var failures = [];
        var typeChecker = program.getTypeChecker();
        var callIdentifiers = tsquery_1.tsquery(sourceFile, "CallExpression Identifier[name=/^(multicast|publish|publishBehavior|publishLast|publishReplay)$/]");
        callIdentifiers.forEach(function (identifier) {
            var callExpression = identifier.parent;
            var fail = false;
            if (identifier.getText() === "multicast") {
                fail = callExpression.arguments.length === 1;
            }
            else {
                fail = !callExpression.arguments.some(function (arg) {
                    return util_1.couldBeFunction(typeChecker.getTypeAtLocation(arg));
                });
            }
            if (fail) {
                failures.push(new Lint.RuleFailure(sourceFile, identifier.getStart(), identifier.getStart() + identifier.getWidth(), Rule.FAILURE_STRING, _this.ruleName));
            }
        });
        return failures;
    };
    Rule.metadata = {
        description: "Disallows operators that return connectable observables.",
        options: null,
        optionsDescription: "Not configurable.",
        requiresTypeInfo: false,
        ruleName: "rxjs-no-connectable",
        type: "functionality",
        typescriptOnly: false
    };
    Rule.FAILURE_STRING = "Connectable observables are forbidden";
    return Rule;
}(Lint.Rules.TypedRule));
exports.Rule = Rule;
