// rxjs-tslint-rules/dist/rules/rxjsNoIgnoredSubscriptionRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Lint = require("tslint");
var tsutils = require("tsutils");
var tsquery_1 = require("@phenomnomnominal/tsquery");
var util_1 = require("../support/util");
var Rule = (function (_super) {
    tslib_1.__extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.applyWithProgram = function (sourceFile, program) {
        var _this = this;
        var failures = [];
        var typeChecker = program.getTypeChecker();
        var callExpressions = tsquery_1.tsquery(sourceFile, "CallExpression[expression.name.text=\"subscribe\"]");
        callExpressions.forEach(function (node) {
            var callExpression = node;
            if (tsutils.isPropertyAccessExpression(callExpression.expression)) {
                if (!tsutils.isExpressionStatement(callExpression.parent)) {
                    return;
                }
                var propertyAccessExpression = callExpression.expression;
                var type = typeChecker.getTypeAtLocation(propertyAccessExpression.expression);
                if (util_1.couldBeType(type, "Observable")) {
                    var name_1 = callExpression.expression.name;
                    failures.push(new Lint.RuleFailure(sourceFile, name_1.getStart(), name_1.getStart() + name_1.getWidth(), Rule.FAILURE_STRING, _this.ruleName));
                }
            }
        });
        return failures;
    };
    Rule.metadata = {
        description: "Disallows ignoring the subscription returned by subscribe.",
        options: null,
        optionsDescription: "Not configurable.",
        requiresTypeInfo: false,
        ruleName: "rxjs-no-ignored-subscription",
        type: "functionality",
        typescriptOnly: false
    };
    Rule.FAILURE_STRING = "Ignoring returned subscriptions is forbidden";
    return Rule;
}(Lint.Rules.TypedRule));
exports.Rule = Rule;
