// angular-alert-module/esm2015/lib/alerts.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AlertsComponent } from './alerts.component';
import { AlertsService } from './alerts.service';
export { AlertsComponent } from './alerts.component';
export { AlertsService } from './alerts.service';
export class AlertsModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: AlertsModule,
            providers: [AlertsService]
        };
    }
}
AlertsModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    AlertsComponent
                ],
                exports: [
                    AlertsComponent
                ]
            },] },
];
function AlertsModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    AlertsModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    AlertsModule.ctorParameters;
}
